bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,488 @@
+extern crate alloc;
+
+use alloc::boxed::Box;
+use alloc::format;
+use alloc::rc::Rc;
+use alloc::string::{String, ToString};
+use alloc::vec;
+use alloc::vec::Vec;
+use core::cell::RefCell;
+use core::cmp::min;
+use core::convert::{From, TryFrom};
+
+#[derive(Debug)]
+pub enum Error {
+    Common(String),
+    TotalSize(String),
+    Header(String),
+    Offset(String),
+    UnknownItem(String),
+    OutOfBound(String),
+    FieldCount(String),
+    Data(String),
+    Overflow(String),
+    Read(String),
+}
+pub trait Read {
+    /**
+     * try to read `buf.len()` bytes from data source with `offset`, then fill it in `buf`.
+     * the return size can be smaller than `buf.len()` which means the remaining data length is
+     * smaller than `buf.len()`
+     */
+    fn read(&self, buf: &mut [u8], offset: usize) -> Result<usize, Error>;
+}
+
+pub const MAX_CACHE_SIZE: usize = 2048;
+pub const MIN_CACHE_SIZE: usize = 64;
+pub const NUM_T_SIZE: usize = 4;
+
+pub struct DataSource {
+    reader: Box<dyn Read>,
+
+    total_size: usize,
+    cache_start_point: usize,
+    // cache size may be smaller than cache.len()
+    cache_size: usize,
+    cache: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub struct Cursor {
```

Comment:
Agree with this, adding my thoughts:

* First of all, I think DataSource and Cursor could benefit from better names. DataSource is simply a block of data that is "in the disk", and DataSource could cache a piece of the data in memory. Cursor is simply a slice, or a viewport into a part of the block of data in DataSource. Cursor, to me here, might not be a proper name name. It might be worthwhile to rename DataSource as well.
* As mentioned above, a huge portion of the code, could be moved to DataSource for better encapsulation between the 2 structures. Think of it like this way: what if someone wants to use DataSource but not Cursor?
* To me, I also find value in customizable cache size. But right now the cache size is pretty fixed, which is a bammer